offline/cable_cru_TRENDY.F90

```diff
@@ -365,104 +380,412 @@ SUBROUTINE CRU_GET_SUBDIURNAL_MET(CRU, MET, CurYear, ktau, kend)
+  !DaysInYear = 365
+  !IF ((CRU%LeapYears) .AND. (is_leapyear(DummyYear))) THEN
+    !DaysInYear = 366
+  !END IF
+
+  !IF (CRU%CTStep == DaysInYear) THEN
+    !! We're at the end of a year
+    !DummyDay = 1
+    !DummyYear = CRU%cYear + 1
+  !END IF
+
+  !! Now we have to handle each of the "next" variables individually
+  !! Tmin
+  !IF (CRU%isRecycled(Tmin)) THEN
+    !DummyYear = CRU%RecycStartYear + MOD(DummyYear - 1501, CRU%MetRecycPeriod)
+  !END IF
+
+  !CALL read_metvals(CRU%MetDatasets(Tmin), CRU%Met(nextTmin)%MetVals, land_x,&
+    !land_y, DummyYear, DummyDay, CRU%LeapYears, CRU%xDimSize, CRU%yDimSize,&
+    !CRU%DirectRead)
+
+  !! vph0900
+  !DummyDay = CRU%CTStep + 1
+  !DummyYear = CRU%cYear
+
+  !! Special handling at the last day of the year
+  !DaysInYear = 365
+  !IF ((CRU%LeapYears) .AND. (is_leapyear(DummyYear))) THEN
+    !DaysInYear = 366
+  !END IF
+
+  !IF (CRU%CTStep == DaysInYear) THEN
+    !! We're at the end of a year
+    !DummyDay = 1
+    !DummyYear = CRU%cYear + 1
+  !END IF
+
+  !IF (CRU%isRecycled(vph0900)) THEN
+    !DummyYear = CRU%RecycStartYear + MOD(DummyYear - 1501, CRU%MetRecycPeriod)
+  !END IF
+
+  !CALL read_metvals(CRU%MetDatasets(vph0900), CRU%Met(nextvph0900)%MetVals,&
+    !land_x, land_y, DummyYear, DummyDay, CRU%LeapYears, CRU%xDimSize,&
+    !CRU%yDimSize, CRU%DirectRead)
+END SUBROUTINE get_daily_met
+
+SUBROUTINE get_met_date(SimYear, SimDay, IsRecycled, RecycleStart,&
+    RecyclePeriod, LeapYears, MetYear, MetDay)
+  !*## Purpose
+  !
+  ! Get the day and year to be used for the meteorology input.
+  !
+  !## Method
+  !
+  ! Use the SimYear and the recycling parameters to determine the date to use
+  ! for the meteorology. The SimDay is checked for validity, as it may be called
+  ! with SimDay+-N at the calling site. If the meteorology is recycled, then
+  ! the year is determined by:
+  ! ```
+  ! RecycleStart + MODULO(SimYear - 1501, RecyclePriod)
+  ! ```
+
+  INTEGER, INTENT(IN) :: SimYear, SimDay, RecycleStart, RecyclePeriod
+  LOGICAL, INTENT(IN) :: IsRecycled, LeapYears
+
+  INTEGER, INTENT(OUT) :: MetYear, MetDay
+
+  ! On leap years and the next/previous day values, and the tricky handling at
+  ! the start and end of recycling periods. The approach for now is to only
+  ! check for recycling at the start of the routine, before doing any year
+  ! shifts when the day = 0 or DaysInYear. This means it is possible to retrieve
+  ! meteorology from outside the recycling period when handling neighbouring
+  ! days.
+  ! For example, we have a recycling period from 1951 to 1960, and we want the
+  ! meteorology from 01/01/1951, and consequently some of the previos day's
+  ! meteorology. The "previous day" with respect to the meteorology can be
+  ! either 31/12/1950 or 31/12/1960, depending on when the MODULO is applied.
+  ! The current approach would use 31/12/1949 as the previous day, which is less
+  ! likely to cause issues with large jumps in the time-local conditions.
```

Comment:
Yep, I did mean 1950. I treat that as being out of the scope of this routine- this just determines what date to send to the ```SpatioTemporalDataset``` reader. Handling edge cases such as this is down to the rules of the ```SpatioTemporalDataset```, which at the moment is to get the date closest to the requested date i.e. the first day of the dataset if asking for a date before the dataset begins, the last for dates after the dataset finishes. 

This was chosen as it was consistent with the existing CRU behaviour. There are two options for handling this I think, either:

1) Current behaviour, get closest date to the requested date (treat 31/12/1950 as 01/01/1951).
2) Treat the year like a modulo and get the requested day from the nearest year (treat 31/12/1950 as 31/12/1951).

I think option 1 is better in the most common instance that this occurs, which is subdiurnalisation. Option 2 is probably better in scenarios where we're operating months/years outside the dataset range, because we'd still get the seasonality, but I don't think this scenario occurs anywhere in CABLE at the moment.